### PR TITLE
13550 verify entity type for sp/gp change and conversion filing

### PR DESCRIPTION
**Pull request by bennypc**
*Issue #:* /bcgov/entity#13550

*Description of changes:*
Created a switch statement to check for the entity type within Change and Conversion UI, checks if entity is a benefit company

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

**Review by JazzarKarim — COMMENTED** [review]
Can you post a screenshot of the console when you try to do a change for BEN for example?

File: src/views/Change.vue [hunk 1]
```diff
@@ -202,6 +203,13 @@ export default class Change extends Mixins(CommonMixin, FeeMixin, FilingTemplate
       // fetch entity snapshot
       const entitySnapshot = await this.fetchEntitySnapshot()
 
+      switch (entitySnapshot.businessInfo.legalType) {
+        case CorpTypeCd.BENEFIT_COMPANY:
+        case CorpTypeCd.BC_COMPANY:
+        case CorpTypeCd.BC_CCC:
+        case CorpTypeCd.BC_ULC_COMPANY:
+          throw new Error(`Invalid entity type, must not be benefit company id`)
+      }
```
Comment (severinbeauvais) [inline src/views/Change.vue:212]: This looks like a good place to check this 👍 

However, this could fail if we add new entity types and forget to check this code. (Why would we? This is for a firm change filing.)

1. I think it would be better to throw the error if the current legal type is NOT a firm (SP or GP).
2. Can `entitySnapshot` be null? Can `entitySnapshow.businessInfo` be null? If so then you can use optional chaining (*) to prevent a de-reference error.

(*) We should use optional chaining sparingly, since it can mask errors that we might want to otherwise see or handle.

**Review by severinbeauvais — COMMENTED** [review]
I think you should also add checks to the other filings. For example:
- Special Resolutions are for coops only
- Alterations are for corps (BC/BEN/CCC/ULC) only
- Limited Restoration Extensions are for corps only
- Limited Restoration to Full are for corps only
- Corp Corrections are for corps only
- Firm Corrections are for firms (SP/GP) only

There will soon be a Special Resolution Correction (or maybe Coop Correction) for coops only -- check with Jia Xu and Travis Semple in RocketChat to see if they add add the entity type check to their code.

**Comment by severinbeauvais** [comment]
Re:
image

Codecov complains when the coverage "goes down". I put that in quotes because the coverage can "go down" even if you make no code changes, so something's not right there. I think we're fine as long as code coverage doesn't drop a lot or go below 80%.

**Comment by bennypc** [comment]
> Can you post a screenshot of the console when you try to do a change for BEN for example?

Screenshot 2023-06-15 at 2 06 39 PM

**Comment by severinbeauvais** [comment]
> Screenshot 2023-06-15 at 2 06 39 PM

Is there an error dialog as well?

Also you should get one of our UX people (Ethan, Yui or Janis) to approve the error message.

File: src/views/Change.vue (hunk 1 above)
Comment (bennypc) [reply src/views/Change.vue:212]: i have implemented both of these changes in the new commit

File: src/views/Change.vue [hunk 2]
```diff
@@ -203,13 +203,14 @@ export default class Change extends Mixins(CommonMixin, FeeMixin, FilingTemplate
...
-        case CorpTypeCd.BC_COMPANY:
-        case CorpTypeCd.BC_CCC:
-        case CorpTypeCd.BC_ULC_COMPANY:
-          throw new Error(`Invalid entity type, must not be benefit company id`)
+      switch (entitySnapshot?.businessInfo?.legalType) {
+        case CorpTypeCd.SOLE_PROP:
+        case CorpTypeCd.PARTNERSHIP:
+          break // acceptable types
+        default:
+          throw new Error(`Invalid entity type, must be a firm (Sole Prop or General Partnership)`)
```
Comment (severinbeauvais) [inline src/views/Change.vue:211]: Can you show a screenshot of this message being displayed please?

Comment (bennypc) [reply src/views/Change.vue:211]: image

File: src/views/Change.vue (hunk 1 above)
Comment (severinbeauvais) [reply src/views/Change.vue:212]: PS Let reviewers resolve the conversations they started -- it's a "bookmark" for them to review their comments. Otherwise they have to re-review the entire code and try to remember what they commented on and see if the code is fixed.

File: src/views/Change.vue (hunk 2 above)
Comment (severinbeauvais) [reply src/views/Change.vue:211]: LGTM

File: tests/unit/CorrectNameOptions.spec.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        propsData: {
+          ...propsData
+        },
+        vuetify
+      })
+    }
+  })
+
+  it('renders the appropriate choices for a numbered company', async () => {
+    correctionNameChoices = ['correct-new-nr']
```
Comment (severinbeauvais) [inline tests/unit/CorrectNameOptions.spec.ts:28]: Might as well use the enum here and below.

File: src/enums/nameChangeOptions.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export enum NameChangeOptions {
+  CORRECT_NAME = 'correct-name',
+  CORRECT_NAME_TO_NUMBER = 'correct-name-to-number',
+  CORRECT_NEW_NR = 'correct-new-nr'
+}
```
Comment (severinbeauvais) [inline src/enums/nameChangeOptions.ts:5]: Don't add this file -- it's already imported from the shared components repo. See the index file in this folder.

File: tests/unit/CorrectNameOptions.spec.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import Vue from 'vue'
```
Comment (severinbeauvais) [inline tests/unit/CorrectNameOptions.spec.ts:1]: Why did you add this test suite? It's not related to the safety checks you implemented...?

Comment (bennypc) [reply tests/unit/CorrectNameOptions.spec.ts:1]: Wait I added it? I don't think I added any test suites

Comment (bennypc) [reply tests/unit/CorrectNameOptions.spec.ts:1]: Should I delete it then?

File: tests/unit/CorrectNameOptions.spec.ts (hunk 3 above)
Comment (severinbeauvais) [reply tests/unit/CorrectNameOptions.spec.ts:28]: Obsolete comment.

File: tests/unit/CorrectNameOptions.spec.ts (hunk 5 above)
Comment (severinbeauvais) [reply tests/unit/CorrectNameOptions.spec.ts:1]: According to GitHub, yes, it's a new file in your commit:

image

If you didn't mean to do this then please delete it in a new commit, and it will disappear from the "total" commits here.

**Comment by bennypc** [comment]
> I think you should also add checks to the other filings. For example:
> 
> * Special Resolutions are for coops only
> * Alterations are for corps (BC/BEN/CCC/ULC) only
> * Limited Restoration Extensions are for corps only
> * Limited Restoration to Full are for corps only
> * Corp Corrections are for corps only
> * Firm Corrections are for firms (SP/GP) only
> 
> There will soon be a Special Resolution Correction (or maybe Coop Correction) for coops only -- check with Jia Xu and Travis Semple in RocketChat to see if they can add the entity type check to their code.

It seems like the other filings already have a way to check this (except for Alterations). For example in Limited Restoration Extension theres

`get restorationResource (): ResourceIF {
    switch (true) {
      case this.isBcCompany: return BcRestorationResource
      case this.isBenefitCompany: return BenRestorationResource
      case this.isBcCcc: return CccRestorationResource
      case this.isBcUlcCompany: return UlcRestorationResource
    }
    return null // should never happen
  }`

**Review by severinbeauvais — APPROVED** [review]
LGTM, but please wait another review or 2 before merging. Thx.

**Review by chenhongjing — APPROVED** [review]
LGTM 😺

**Review by JazzarKarim — APPROVED** [review]
LGTM 👍
